File: week1/Ariling/boj-2470-두 용액/main.py
import sys
input = sys.stdin.readline

# 이분 탐색
inf = int(1e9)
N = int(input())
arr = list(map(int, input().split()))
arr = sorted(arr)
left , right = 0, N-1
result = [arr[left], arr[right]]
min_dif = abs(arr[left] + arr[right])
while left < right:
    dif = arr[left] + arr[right]
    if abs(dif) < min_dif:
        min_dif = abs(dif)
        result = [arr[left], arr[right]]
        if min_dif == 0:
            break
    if dif < 0:
        left += 1
    else:
        right -= 1
print(result[0], result[1])

File: week1/Ariling/boj-2470-두 용액/main.js
// 문제 풀이 함수
function solution(input) {
  input.sort((a, b) => a - b);
  (st = 0), (ed = input.length - 1);
  let result = [input[st], input[ed]];
  let diff = Math.abs(result[0] + result[1]);
  while (st < ed) {
    let cur_diff = input[st] + input[ed];
    if (Math.abs(cur_diff) < diff) {
      diff = Math.abs(cur_diff);
      result = [input[st], input[ed]];
      if (diff === 0) {
        break;
      }
    }
    if (cur_diff < 0) {
      st++;
    } else {
      ed--;
    }
  }
  console.log(result[0], result[1]);
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n;
let arr;

rl.on("line", (line) => {
  if (!n) {
    n = parseInt(line);
  } else {
    arr = line.split(" ").map((el) => {
      return Number(el);
    });
    rl.close();
  }
}).on("close", () => {
  solution(arr);
  process.exit();
});
